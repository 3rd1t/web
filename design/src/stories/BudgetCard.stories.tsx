import React from "react";
import { BudgetCard } from '../components/BudgetCard';

export default {
  title: "Budget Card",
};

export const card = () => {

  return (
    <BudgetCard  expensesCount= {195} totalBudget={1500.00} start={new Date()} end={new Date()} budgeted={4.99} expenses={625.12} income={10.00}/>
  )
};
