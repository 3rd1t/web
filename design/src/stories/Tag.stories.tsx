import React from 'react';
import Tag from "../components/Tag"
export default { title: 'Tag' };

export const Default = () => {
    return (
        <Tag>Hello Tag</Tag>
    )
} 
