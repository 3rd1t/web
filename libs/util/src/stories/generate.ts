/**
 * Return all components from storybook.
 *
 * @example
 * import * as stories from "./<component>.stories"
 *
 * const testCases = generateTestCasesFromStories(stories)
 * testCases.forEach((s) => {
 *   it(s.name + "should match snapshot", () => {
 *     const { baseElement } = render(s.component())
 *     expect(baseElement).toMatchSnapshot()
 *   })
 * })
 *
 *
 * @param stories - import * as stories from "./<component>.stories"
 * @returns Name and render function for each component
 */
export function generateTestCasesFromStories(
  stories: Record<string, any>, // eslint-disable-line
): { name: string; component: () => HTMLElement }[] {
  return Object.entries(stories)
    .filter((s) => s[0] !== "default")
    .map((s) => {
      return {
        name: s[0],
        component: s[1],
      }
    })
}
