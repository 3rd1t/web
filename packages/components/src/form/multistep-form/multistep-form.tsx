import React, { useEffect, useState } from "react"
import { Select } from "@perfolio/components/form/select/select"
import { AnimatePresence, AnimateSharedLayout, motion } from "framer-motion"
import { Dots } from "../../progress/dots/dots"

type Step = {
  name: string
  description: string
  choices: string[]
  choice: number
  setChoice: React.Dispatch<React.SetStateAction<number>>
}

/* eslint-disable-next-line */
export interface MultistepFormProps {
  title: string
  steps: Step[]
}

export const MultistepForm = (props: MultistepFormProps) => {
  const [progress, setProgress] = useState<number>(0)
  const [active, setActive] = useState<number>(0)

  const updateState = (step: Step) => (newChoice: number) => {
    // Only advance if the field did not have a selection in the first place.
    if (step.choice < 0) {
      setProgress(progress + 1)
    }
    setActive(progress + 1)
    step.setChoice(newChoice)
  }
  useEffect(() => {
    setActive(progress)
  }, [progress])

  return (
    <>
      <div className="flex items-center justify-between">
        <h3 className="text-lg font-semibold text-gray-800">Build your custom file</h3>
        <Dots current={progress} max={props.steps.length} />
      </div>

      <div className="space-y-4">
        <AnimateSharedLayout>
          {props.steps.map((step, i) => {
            return (
              <AnimatePresence key={i}>
                {i <= progress ? (
                  <motion.div
                    layout
                    initial={{ opacity: 0, y: 20 }}
                    animate={{ opacity: 1, y: 0 }}
                    transition={{
                      damping: 300,
                    }}
                    onClick={() => {
                      if (active !== i) {
                        setActive(i)
                      }
                    }}
                    className="flex flex-col p-3 bg-white border border-gray-300 rounded md:flex-row md:items-center md:justify-between hover:border-data-500"
                  >
                    <div className="md:w-2/3">
                      <div className="flex items-center justify-start space-x-3">
                        <AnimatePresence>
                          {i < progress && i !== active ? (
                            <motion.div
                              layout
                              initial={{ opacity: 0, x: -20, width: 0 }}
                              animate={{ opacity: 1, x: 0, width: "auto" }}
                              exit={{ opacity: 0, x: -20, width: 0 }}
                              transition={{ damping: 300 }}
                            >
                              <svg
                                className="w-6 h-6 text-data-600"
                                fill="currentColor"
                                viewBox="0 0 20 20"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  fillRule="evenodd"
                                  d="M16.707 5.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L8 12.586l7.293-7.293a1 1 0 011.414 0z"
                                  clipRule="evenodd"
                                />
                              </svg>
                            </motion.div>
                          ) : null}
                        </AnimatePresence>
                        <motion.h3 layout className="text-2xl font-extrabold text-gray-900">
                          {step.name}
                        </motion.h3>
                      </div>
                      <AnimatePresence>
                        {i === progress || i === active ? (
                          <motion.p
                            layout
                            initial={{ opacity: 0, height: 0, y: -10 }}
                            animate={{ opacity: 1, height: "auto", y: 0 }}
                            exit={{ opacity: 0, height: 0, y: -10 }}
                            transition={{ damping: 400 }}
                            className="text-sm text-gray-700 "
                          >
                            {step.description}
                          </motion.p>
                        ) : null}
                      </AnimatePresence>
                    </div>

                    <div className="w-full mt-4 md:w-1/3 md:mt-0">
                      <Select
                        label={step.name}
                        selected={step.choice}
                        setSelected={updateState(step)}
                        choices={step.choices}
                      />
                    </div>
                  </motion.div>
                ) : null}
              </AnimatePresence>
            )
          })}
        </AnimateSharedLayout>
      </div>
    </>
  )
}

export default MultistepForm
