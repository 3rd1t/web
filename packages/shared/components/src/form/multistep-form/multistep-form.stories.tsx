import React from "react"
import { MultistepForm } from "./multistep-form"

export default {
  component: MultistepForm,
  title: "components/Form/MultistepForm",
}

export const twoSteps = () => {
  const first = -1
  const setFirst = (v: number) => {}
  const second = -1
  const setSecond = (v: number) => {}

  const props: {
    steps: {
      name: string
      description: string
      choices: string[]
      value: number
      setValue: React.Dispatch<React.SetStateAction<number>>
    }[]
  } = {
    steps: [
      {
        name: "First Question",
        description: "Description",
        choices: ["Hello", "World"],
        value: first,
        setValue: setFirst,
      },
      {
        name: "Second Question",
        description: "Description",
        choices: ["Hello", "World"],
        value: second,
        setValue: setSecond,
      },
    ],
  }

  return <MultistepForm steps={props.steps} />
}

export const realExample = () => {
  /* eslint-disable-next-line */
  const props = {
    factorModels: [
      {
        name: "3factor",
        description: "3 Factor model by Fama & French 1993",
        factors: [
          {
            name: "XXX",
            description: "maybe just spell it out",
          },
          {
            name: "XXX",
            description: "maybe just spell it out",
          },
          {
            name: "XXX",
            description: "maybe just spell it out",
          },
          {
            name: "all",
            description: "All factors",
          },
        ],
      },
      {
        name: "4factor",
        description: "4 Factor model by Fama & French 1993",

        factors: [
          {
            name: "XXX",
            description: "maybe just spell it out",
          },
          {
            name: "XXX",
            description: "maybe just spell it out",
          },
          {
            name: "XXX",
            description: "maybe just spell it out",
          },
          {
            name: "XXX",
            description: "maybe just spell it out",
          },
          {
            name: "all",
            description: "All factors",
          },
        ],
      },
      {
        name: "5factor",
        description: "5 Factor model by Fama & French 1993",

        factors: [
          {
            name: "XXX",
            description: "maybe just spell it out",
          },
          {
            name: "XXX",
            description: "maybe just spell it out",
          },
          {
            name: "XXX",
            description: "maybe just spell it out",
          },
          {
            name: "XXX",
            description: "maybe just spell it out",
          },
          {
            name: "XXX",
            description: "maybe just spell it out",
          },

          {
            name: "all",
            description: "All factors",
          },
        ],
      },
      {
        name: "6factor",
        description: "6 Factor model by Fama & French 1993",

        factors: [
          {
            name: "mom",
            description: "Momentum",
          },
          {
            name: "XXX",
            description: "maybe just spell it out",
          },
          {
            name: "XXX",
            description: "maybe just spell it out",
          },
          {
            name: "XXX",
            description: "maybe just spell it out",
          },
          {
            name: "XXX",
            description: "maybe just spell it out",
          },
          {
            name: "XXX",
            description: "maybe just spell it out",
          },

          {
            key: "all",
            description: "All factors",
          },
        ],
      },
    ],
    regions: [
      "USA",
      "Developed",
      "Developed_ex_US",
      "Europe",
      "Japan",
      "Asia_Pacific_ex_Japan",
      "North_America",
      "Emerging",
    ],

    currencies: [
      "EUR",
      "JPY",
      "GBP",
      "CHF",
      "RUB",
      "AUD",
      "BRL",
      "CAD",
      "CNY",
      "INR",
      "DKK",
      "NZD",
      "NOK",
      "SEK",
      "PLN",
      "ILS",
      "KRW",
      "TRY",
    ],

    intervals: ["daily", "monthly", "annual"],
  }

  const model = -1
  const setModel = (v: number) => {}
  const factor = -1
  const setFactor = (v: number) => {}

  const region = -1
  const setRegion = (v: number) => {}
  const currency = -1
  const setCurrency = (v: number) => {}
  const interval = -1
  const setInterval = (v: number) => {}

  const steps: {
    name: string
    description: string
    choices: string[]
    value: number
    setValue: React.Dispatch<React.SetStateAction<number>>
  }[] = [
    {
      name: "Model",
      description: "Quisque ultrices odio ut tellus congue, scelerisque egestas augue accumsan.",
      choices: props.factorModels.map((m) => m.description),
      value: model,
      setValue: setModel,
    },
    {
      name: "Factor",
      description: "Quisque ultrices odio ut tellus congue, scelerisque egestas augue accumsan.",
      choices:
        model >= 0
          ? props.factorModels.find((f) => f.name === props.factorModels[model].name).factors.map((f) => f.description)
          : [],
      value: factor,
      setValue: setFactor,
    },
    {
      name: "Region",
      description: "Quisque ultrices odio ut tellus congue, scelerisque egestas augue accumsan.",
      choices: props.regions,
      value: region,
      setValue: setRegion,
    },
    {
      name: "Currency",
      description: "Quisque ultrices odio ut tellus congue, scelerisque egestas augue accumsan.",
      choices: props.currencies,
      value: currency,
      setValue: setCurrency,
    },
    {
      name: "Interval",
      description: "Quisque ultrices odio ut tellus congue, scelerisque egestas augue accumsan.",
      choices: props.intervals,
      value: interval,
      setValue: setInterval,
    },
  ]

  return <MultistepForm steps={steps} />
}
