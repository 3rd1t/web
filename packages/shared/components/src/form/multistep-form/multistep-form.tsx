import React, { useEffect, useState } from "react"
import { Select } from "@perfolio/shared/components"
import { AnimatePresence, AnimateSharedLayout, motion } from "framer-motion"

type Step = {
  name: string
  description: string
  choices: string[]
  value: number
  setValue: React.Dispatch<React.SetStateAction<number>>
}

/* eslint-disable-next-line */
export interface MultistepFormProps {
  steps: Step[]
}

export const MultistepForm = (props: MultistepFormProps) => {
  const [progress, setProgress] = useState<number>(0)
  const [active, setActive] = useState<number>(0)

  const updateState = (step: Step) => (newValue: number) => {
    console.log({ progress, active })
    // Only advance if the field did not have a selection in the first place.
    if (step.value < 0) {
      setProgress(progress + 1)
    }
    console.log("running")
    setActive(progress + 1)
    console.log("ran")

    step.setValue(newValue)
  }

  useEffect(() => {
    console.log({ active })
  }, [active])

  useEffect(() => {
    setActive(progress)
  }, [progress])

  return (
    <div className="space-y-4">
      <AnimateSharedLayout>
        {props.steps.map((step, i) => {
          return (
            <AnimatePresence>
              {i <= progress ? (
                <motion.div
                  layout
                  initial={{ opacity: 0, y: 20 }}
                  animate={{ opacity: 1, y: 0 }}
                  transition={{
                    damping: 300,
                  }}
                  onClick={() => {
                    if (active !== i) {
                      setActive(i)
                    }
                  }}
                  className={`flex flex-col md:flex-row  md:items-center p-3 border md:justify-between border-gray-400 hover:bg-gray-100 rounded ${
                    i === progress || i === active ? "bg-gray-100" : ""
                  }`}
                >
                  <div className="md:w-2/3">
                    <div className="flex items-center justify-start space-x-3">
                      <AnimatePresence>
                        {i < progress && i !== active ? (
                          <motion.div
                            layout
                            initial={{ opacity: 0, x: -20, width: 0 }}
                            animate={{ opacity: 1, x: 0, width: "auto" }}
                            exit={{ opacity: 0, x:-20, width: 0 }}
                            transition={{ damping: 300 }}
                          >
                            <svg
                              className="w-6 h-6 text-data-600"
                              fill="currentColor"
                              viewBox="0 0 20 20"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                fillRule="evenodd"
                                d="M16.707 5.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L8 12.586l7.293-7.293a1 1 0 011.414 0z"
                                clipRule="evenodd"
                              />
                            </svg>
                          </motion.div>
                        ) : null}
                      </AnimatePresence>
                      <motion.h3 layout className="text-2xl font-extrabold text-gray-900">
                        {step.name}
                      </motion.h3>
                    </div>
                    <AnimatePresence>
                      {i === progress || i === active ? (
                        <motion.p
                          layout
                          initial={{ opacity: 0, height: 0, y: -10 }}
                          animate={{ opacity: 1, height: "auto", y: 0 }}
                          exit={{ opacity: 0, height: 0, y: -10 }}
                          transition={{ damping: 400 }}
                          className="text-sm text-gray-500 "
                        >
                          {step.description}
                        </motion.p>
                      ) : null}
                    </AnimatePresence>
                  </div>

                  <div className="w-full mt-4 md:w-1/3 md:mt-0">
                    <Select selected={step.value} setSelected={updateState(step)} choices={step.choices} />
                  </div>
                </motion.div>
              ) : null}
            </AnimatePresence>
          )
        })}
      </AnimateSharedLayout>
    </div>
  )
}

export default MultistepForm
