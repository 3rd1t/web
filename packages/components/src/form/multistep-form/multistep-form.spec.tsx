import { render } from "@testing-library/react"
import * as stories from "./multistep-form.stories"
import { generateTestCasesFromStories } from "@perfolio/util/stories/generate"

describe("MultistepForm", () => {
  describe("Stories", () => {
    const testCases = generateTestCasesFromStories(stories)
    testCases.forEach((s) => {
      it(s.name + "should match snapshot", () => {
        const { baseElement } = render(s.component())
        expect(baseElement).toMatchSnapshot()
      })
    })
  })
})
