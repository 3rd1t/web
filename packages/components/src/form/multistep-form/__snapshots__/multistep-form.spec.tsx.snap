// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultistepForm Stories realExampleshould match snapshot 1`] = `
<body>
  <div>
    <div
      class="flex items-center justify-between"
    >
      <h3
        class="text-lg font-semibold text-gray-800"
      >
        Build your custom file
      </h3>
      <div
        class="flex items-center space-x-6"
      >
        <span
          class=" rounded-full bg-data-700 border-4 w-4 h-4 border-data-400"
        />
        <span
          class=" rounded-full w-2 h-2 bg-gray-400"
        />
        <span
          class=" rounded-full w-2 h-2 bg-gray-400"
        />
        <span
          class=" rounded-full w-2 h-2 bg-gray-400"
        />
        <span
          class=" rounded-full w-2 h-2 bg-gray-400"
        />
      </div>
    </div>
    <div
      class="space-y-4"
    >
      <div
        class="flex flex-col p-3 bg-white border border-gray-300 rounded md:flex-row md:items-center md:justify-between hover:border-data-500"
        style="opacity: 0; transform: translateY(20px) translateZ(0);"
      >
        <div
          class="md:w-2/3"
        >
          <div
            class="flex items-center justify-start space-x-3"
          >
            <h3
              class="text-2xl font-extrabold text-gray-900"
            >
              Model
            </h3>
          </div>
          <p
            class="text-sm text-gray-700 "
            style="opacity: 0; height: 0px; transform: translateY(-10px) translateZ(0);"
          >
            Quisque ultrices odio ut tellus congue, scelerisque egestas augue accumsan.
          </p>
        </div>
        <div
          class="w-full mt-4 md:w-1/3 md:mt-0"
        >
          <div
            class="flex items-center justify-center"
          >
            <div
              class="w-full max-w-xs mx-auto"
            >
              <div
                class="space-y-1"
              >
                <div
                  class="relative"
                >
                  <span
                    class="inline-block w-full rounded-md shadow-sm"
                  >
                    <button
                      aria-haspopup="true"
                      aria-label="Model"
                      class="relative w-full py-2 pl-3 pr-10 text-left transition duration-150 ease-in-out bg-white border border-gray-300 rounded-md cursor-default focus:outline-none focus:shadow-outline-blue focus:border-blue-300 sm:text-sm sm:leading-5"
                      id="headlessui-listbox-button-2"
                      type="button"
                    >
                      <span
                        class="flex items-center h-5 truncate"
                      />
                      <span
                        class="absolute inset-y-0 right-0 flex items-center pr-2 pointer-events-none"
                      >
                        <svg
                          class="w-5 h-5 text-gray-400"
                          fill="none"
                          stroke="currentColor"
                          viewBox="0 0 20 20"
                        >
                          <path
                            d="M7 7l3-3 3 3m0 6l-3 3-3-3"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="1.5"
                          />
                        </svg>
                      </span>
                    </button>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`MultistepForm Stories twoStepsshould match snapshot 1`] = `
<body>
  <div>
    <div
      class="flex items-center justify-between"
    >
      <h3
        class="text-lg font-semibold text-gray-800"
      >
        Build your custom file
      </h3>
      <div
        class="flex items-center space-x-6"
      >
        <span
          class=" rounded-full bg-data-700 border-4 w-4 h-4 border-data-400"
        />
        <span
          class=" rounded-full w-2 h-2 bg-gray-400"
        />
      </div>
    </div>
    <div
      class="space-y-4"
    >
      <div
        class="flex flex-col p-3 bg-white border border-gray-300 rounded md:flex-row md:items-center md:justify-between hover:border-data-500"
        style="opacity: 0; transform: translateY(20px) translateZ(0);"
      >
        <div
          class="md:w-2/3"
        >
          <div
            class="flex items-center justify-start space-x-3"
          >
            <h3
              class="text-2xl font-extrabold text-gray-900"
            >
              First Question
            </h3>
          </div>
          <p
            class="text-sm text-gray-700 "
            style="opacity: 0; height: 0px; transform: translateY(-10px) translateZ(0);"
          >
            Description
          </p>
        </div>
        <div
          class="w-full mt-4 md:w-1/3 md:mt-0"
        >
          <div
            class="flex items-center justify-center"
          >
            <div
              class="w-full max-w-xs mx-auto"
            >
              <div
                class="space-y-1"
              >
                <div
                  class="relative"
                >
                  <span
                    class="inline-block w-full rounded-md shadow-sm"
                  >
                    <button
                      aria-haspopup="true"
                      aria-label="First Question"
                      class="relative w-full py-2 pl-3 pr-10 text-left transition duration-150 ease-in-out bg-white border border-gray-300 rounded-md cursor-default focus:outline-none focus:shadow-outline-blue focus:border-blue-300 sm:text-sm sm:leading-5"
                      id="headlessui-listbox-button-1"
                      type="button"
                    >
                      <span
                        class="flex items-center h-5 truncate"
                      />
                      <span
                        class="absolute inset-y-0 right-0 flex items-center pr-2 pointer-events-none"
                      >
                        <svg
                          class="w-5 h-5 text-gray-400"
                          fill="none"
                          stroke="currentColor"
                          viewBox="0 0 20 20"
                        >
                          <path
                            d="M7 7l3-3 3 3m0 6l-3 3-3-3"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="1.5"
                          />
                        </svg>
                      </span>
                    </button>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
